components/playlist/browser/playlist_download_request_manager.cc

```diff
@@ -329,43 +238,120 @@ void PlaylistDownloadRequestManager::ProcessFoundMedia(
       item->thumbnail_source = GURL(*thumbnail);
       item->thumbnail_path = GURL(*thumbnail);
     }
-    item->media_source = GURL(*src);
-    item->media_path = GURL(*src);
-    if (author)
+    if (author) {
       item->author = *author;
-
+    }
     items.push_back(std::move(item));
   }
 
-  DVLOG(2) << __func__
-           << " Media detection result size: " << value.GetList().size() << " "
-           << items.size();
-  std::move(callback).Run(std::move(items));
+  DVLOG(2) << __func__ << " Media detection result size: " << items.size();
+
+  if (contents == background_contents() && items.size()) {
+    CHECK(!callback_for_current_request_.is_null()) << " callback already ran";
+    auto callback = std::move(callback_for_current_request_);
+
+    DCHECK_GT(in_progress_urls_count_, 0);
+    in_progress_urls_count_--;
+
+    std::vector<mojom::PlaylistItemPtr> cloned_items;
+    base::ranges::transform(items, std::back_inserter(cloned_items),
+                            &mojom::PlaylistItemPtr::Clone);
+    std::move(callback).Run(std::move(cloned_items));
+
+    web_contents_.reset();
+  }
+
+  return items;
 }
 
+bool PlaylistDownloadRequestManager::CanCacheMedia(
+    const mojom::PlaylistItemPtr& item) const {
+  GURL media_url(item->media_source);
+  if (media_url.SchemeIs(url::kHttpsScheme)) {
+    return true;
+  }
+
+  if (media_url.SchemeIsBlob()) {
+    if (item->is_blob_from_media_source) {
+      // At this moment, we have a few sites that we can get media files with
+      // hacks.
+      return media_detector_component_manager_->ShouldHideMediaSrcAPI(
+                 media_url) ||
+             media_detector_component_manager_->ShouldUseFakeUA(media_url);
+    }
+
+    // blob: which is not Media Source
+    // TODO(sko) Test and allow this case referring to
+    // https://github.com/brave/brave-core/pull/17246
+    return false;
+  }
+
+  return false;
+}
+
+bool PlaylistDownloadRequestManager::ShouldRefetchMediaSourceToCache(
+    const mojom::PlaylistItemPtr& item) const {
+  GURL media_url(item->media_source);
+  if (media_url.SchemeIs(url::kHttpsScheme)) {
+    return false;
```

Comment:
Not sure I understand the logic here. Doesn't `media_source` always point to the original media URL? Shouldn't it return `true` here?

Reply:
This method is more like "We should refetch the media source url **using background web contents** when the item's media source url is blob:".

So in case scheme is non blob, "https" scheme we don't have to refetch as we can cache the media with the url as is.

Reply:
Should we rename this function so that this is more apparent then? E.g. `ShouldExtractMediaFromOffscreenWebContents()`?